servo/connectors/newrelic.py

```diff
@@ -0,0 +1,340 @@
+                When None, all configured metrics are measured. Defaults to None.
+            control (Control, optional): A control descriptor that describes how            
+                the measurement is to be captured. Defaults to Control().
+
+        Returns:
+            Measurement: An object that aggregates the state of the metrics
+            queried from Newrelic.
+        """
+        if metrics:
+            metrics__ = list(filter(lambda m: m.name in metrics, self.metrics))
+        else:
+            metrics__ = self.metrics
+        measuring_names = list(map(lambda m: m.name, metrics__))
+        self.logger.info(f"Starting measurement of {len(metrics__)} metrics: {join_to_series(measuring_names)}")
+
+        start = datetime.now() + control.warmup
+        end = start + control.duration
+
+        sleep_duration = Duration(control.warmup + control.duration)
+        self.logger.info(
+            f"Waiting {sleep_duration} during metrics collection ({control.warmup} warmup + {control.duration} duration)..."
+        )
+
+        progress = DurationProgress(sleep_duration)
+        notifier = lambda p: self.logger.info(p.annotate(f"waiting {sleep_duration} during metrics collection...", False), progress=p.progress)
+        await progress.watch(notifier)
+        self.logger.info(f"Done waiting {sleep_duration} for metrics collection, resuming optimization.")
+
+        # Fetch the measurements
+        self.logger.info(f"Querying Newrelic for {len(metrics__)} metrics...")
+        readings = await self._query_newrelic(metrics__, start, end)
+        measurement = Measurement(readings=readings)
+        return measurement
+
+    async def _query_newrelic(
+        self, metrics: List[NewrelicMetric], start: datetime, end: datetime
+    ) -> List[TimeSeries]:
+        f_names_to_ms: Dict[str, List[NewrelicMetric]] = defaultdict(list)
+        fetches: Dict[str, Set[str]] = defaultdict(set)
+        for m in metrics:
+            f_names_to_ms[m.fetch_name].append(m)
+            fetches[m.fetch_name].add(m.values_selector)
+
+        newrelic_request = NewrelicRequest(base_url=self.config.api_url, fetches=fetches, start=start, end=end, step=self.config.step)
+
+        self.logger.trace(f"Getting Newrelic instance ids: {newrelic_request.url}")
+        instance_ids: List[str] = await self.dispatch_event("instances")
+        
+        self.logger.trace(f"Querying Newrelic: {newrelic_request.url}")
+        readings = []
+        # TODO asyncio gather this instead
+        for i in instance_ids:
+            api_path = '/applications/{app_id}/instances/{instance_id}/metrics/data.json'.format(NEWRELIC_APM_APP_ID, i)
+            self.logger.trace(f"Querying Newrelic for instance: {i}")
+            async with httpx.AsyncClient(
+                base_url=self.config.api_url + api_path, 
+                params=newrelic_request.params,
+                headers={'X-Api-Key': NEWRELIC_APM_API_KEY},
```

Comment:
`NEWRELIC_APM_API_KEY` must come from config to support multiple New Relic connector instances